GPT/gpt.talon

```diff
@@ -36,6 +36,19 @@ model apply [from] clip$:
     result = user.gpt_apply_prompt(prompt, text)
     user.paste(result)
 
+model context add [{user.modelSource}]:
+    text = user.gpt_get_source_text(modelSource or "")
+    user.gpt_push_context(text)
+model context clear: user.gpt_clear_context()
+model context optimize: user.gpt_optimize_context()
+
+model thread <user.modelPrompt> [{user.modelSource}] [{user.modelDestination}]:
+    text = user.gpt_get_source_text(modelSource or "")
+    result = user.gpt_apply_prompt(modelPrompt, text, "thread")
+    user.gpt_insert_response(result, modelDestination or "", "thread")
+model thread new: user.gpt_new_thread()
+model thread optimize: user.gpt_optimize_thread()
```

Comment:
I am not very clear on how thread is different from normal context. So this puts the previous response in the context right?

I feel that for users this might be confusing. I wonder if we can use a different term.

Reply:
Thread and context are separate-context will be sent with every request regardless. Thread conversations will continue only if you use the thread command. This lets you set up the some baseline context that you can use for all conversations (think custom gpt) that you curate. That context is valuable since you spent the time to pick it out specifically. The thread content is specific to one conversation and you probably want to get rid of it when you start the next topic.

Reply:
This is cool and makes sense. I think it is a good idea. I think we need to think about ways to manage state with this however. I think it is going to be confusing for a user to know what is in the thread or how to use these.

I think thread is fine as a keyword. It is shorter than others I can think of. I was thinking of threading in software not threads on a chat app. 

I am thinking we might want to have some sort sort of visual indicator that shows the number of lines / first few words in the stored context or something analogous. Just so it is very clear what is being sent/stored. And then we can use the same indicator when a request is in process or something . Does that seem reasonable?

Reply:
I like that idea but I'm not sure how to implement it. Can you contribute that functionality? If it seems challenging, we could also plan to to it as a follow up. This is really useful as-is, and the confusion isn't too bad in my experience since it's quite obvious from the responses you get that there is more in the context than you thought. I've been using this branch all week.

Reply:
As an mvp, we could make "thread" and "context" into sources that can be "passed"...?

Reply:
Now that I'm thinking of it, context might make sense as a destination too, as an alternative to the custom grammar "model context add", you could "model pass to context". Though you'd still need something custom to "clear" it. I'm not sure if that's better.

Reply:
I played with this a little bit and it's not very straightforward to fit it into the existing grammar because I'm not sure how to change the code to fit it in. I'm still finding it quite useful as is.

Reply:
I think I found a middle ground that adopts some grammar from the existing pass command.

Reply:
Yes do not worry about an indicator right now. That is some thing I would like to add eventually but we don't need to do now. 

For this because it is not really doing a model operation to my understanding but only getting a model operation ready, I wouldn't worry too much about fitting it into the grammar. Do whatever is best for you